### PR TITLE
1603 by Inlead: Implement ding_dashboard into DDB CMS.

### PR DESCRIPTION
**Pull request by inlead**
http://platform.dandigbib.org/issues/1603

### CODE REVIEW

**Comment by cableman** [comment]
Please write the README file with information about what this module do and how to configure it.

File: modules/ding_dashboard/ding_dashboard.admin.inc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ */
+function ding_dashboard_admin_content_types($form, $form_state) {
+  $node_types = node_type_get_types();
+  $node_bundles = array();
+
+  foreach ($node_types as $type) {
+    $node_bundles[$type->type] = $type->name;
+  }
+
+  $form['editorial_content_types'] = array(
```
Comment (cableman) [inline modules/ding_dashboard/ding_dashboard.admin.inc]: Please prefix both the form indexes and the variables store in "variables" table with module name (ding_).

File: modules/ding_dashboard/ding_dashboard.admin.inc [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  $node_types = array_filter($input['editorial_content_types']);
+  $last_setting = array_filter(variable_get('editorial_content_types', array()));
+
+  if (count($node_types) == 0) {
+    return;
+  }
+
+  $instance_definition = array(
+    'label' => 'Editorial status',
+    'field_name' => 'field_editorial_status',
```
Comment (cableman) [inline modules/ding_dashboard/ding_dashboard.admin.inc]: Prefix fields to contain the module name (field_ding_).

**Comment by cableman** [comment]
As I don't have the README: It seam to me that this PR do the same stuff ass https://www.drupal.org/project/workbench_moderation and workbench can be configured to do.

File: modules/ding_dashboard/ding_dashboard.module [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+    'description' => t('Allow to delete any editorial dashboard content.'),
+  );
+
+  return $perm;
+}
+
+/**
+ * Implements hook_secure_permissions().
+ */
+function ding_dashboard_secure_permissions($role) {
```
Comment (cableman) [inline modules/ding_dashboard/ding_dashboard.module]: Permission should be moved to ding_permissions.

File: modules/ding_dashboard/ding_dashboard.module [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+ *   Action name.
+ *
+ * @return array
+ *   Returns an array of permissions.
+ */
+function ding_dashboard_access_callback($node, $action) {
+  global $user;
+  $perms = &drupal_static(__FUNCTION__);
+  if (!isset($perms[$user->uid][$node->nid][$action])) {
+    switch ($action) {
```
Comment (cableman) [inline modules/ding_dashboard/ding_dashboard.module]: It seams that the check "$user->uid == $node->uid" could be move outside the switch as a fast check.

File: modules/ding_dashboard/ding_dashboard.module [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+function ding_dashboard_node_access($node, $op, $account) {
+  $editorial_types = array_filter(variable_get('editorial_content_types', array()));
+  $state = NODE_ACCESS_IGNORE;
+  $ed_states = array(EDITORIAL_NODE_APPROVED, EDITORIAL_NODE_PUSHED);
+
+  if (is_object($node) && in_array($node->type, $editorial_types)) {
+    switch ($op) {
+      case 'update':
+        $ed_status = $node->field_editorial_status[LANGUAGE_NONE][0]['value'];
+        if (in_array($ed_status, $ed_states) && in_array('guest blogger', $account->roles) && $node->uid == $account->uid) {
```
Comment (cableman) [inline modules/ding_dashboard/ding_dashboard.module]: The check "$node->uid == $account->uid" could be first as a fast check.

File: modules/ding_dashboard/modules/ding_dashboard_notification/lib/class.phpmailer.php [hunk 6]
```diff
@@ -0,0 +1,2397 @@
+<?php
```
Comment (cableman) [inline modules/ding_dashboard/modules/ding_dashboard_notification/lib/class.phpmailer.php]: Third party libraries should not be part of the module.

Why is PHPMailer needed? An if it is it should be loaded using the libraries module.

**Comment by DDBdrift** [comment]
https://platform.dandigbib.org/issues/1603 | Closed

**Pull request closed without merging**